hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/TimeoutOptimizer.java

```diff
@@ -0,0 +1,227 @@
+    private int connTimeout = -1;
+    private int maxReqTimeout;
+    private int timeoutIncRate;
+    private boolean shouldOptimizeTimeout;
+
+    public TimeoutOptimizer(URL url, AbfsRestOperationType opType, ExponentialRetryPolicy retryPolicy, AbfsConfiguration abfsConfiguration) {
+        this.url = url;
+        this.opType = opType;
+        if (opType != null) {
+            this.retryPolicy = retryPolicy;
+            this.abfsConfiguration = abfsConfiguration;
+            if (abfsConfiguration.get(ConfigurationKeys.AZURE_OPTIMIZE_TIMEOUTS) == null) {
+                this.shouldOptimizeTimeout = false;
+            }
+            else {
+                this.shouldOptimizeTimeout = Boolean.parseBoolean(abfsConfiguration.get(ConfigurationKeys.AZURE_OPTIMIZE_TIMEOUTS));
+            }
+            if (this.shouldOptimizeTimeout) {
+                this.maxReqTimeout = Integer.parseInt(abfsConfiguration.get(ConfigurationKeys.AZURE_MAX_REQUEST_TIMEOUT));
+                this.timeoutIncRate = Integer.parseInt(abfsConfiguration.get(ConfigurationKeys.AZURE_REQUEST_TIMEOUT_INCREASE_RATE));
+                initTimeouts();
+                updateUrl();
+            }
+
+        } else {
+            this.shouldOptimizeTimeout = false;
+        }
+    }
+
+    public void updateRetryTimeout(int retryCount) {
+        if (!this.shouldOptimizeTimeout) {
+            return;
+        }
+
+        // update all timeout values
+        updateTimeouts(retryCount);
+        updateUrl();
+    }
+
+    public URL getUrl() {
+        return url;
+    }
+    public boolean getShouldOptimizeTimeout() { return this.shouldOptimizeTimeout; }
+
+    public int getRequestTimeout() { return requestTimeout; }
+
+    public int getReadTimeout() {
+        return readTimeout;
+    }
+
+    public int getReadTimeout(final int defaultTimeout) {
+        if (readTimeout != -1 && shouldOptimizeTimeout) {
+            return readTimeout;
+        }
+        return defaultTimeout;
+    }
+
+    public int getConnTimeout() {
+        return connTimeout;
+    }
+
+    public int getConnTimeout(final int defaultTimeout) {
+        if (connTimeout == -1) {
+            return defaultTimeout;
+        }
+        return connTimeout;
+    }
+
+    private void initTimeouts() {
+        if (!shouldOptimizeTimeout) {
+            requestTimeout = -1;
+            readTimeout = -1;
+            connTimeout = -1;
+            return;
+        }
+
+        String query = url.getQuery();
+        int timeoutPos = query.indexOf("timeout");
+        if (timeoutPos < 0) {
+            // no value of timeout exists in the URL
+            // no optimization is needed for this particular request as well
+            requestTimeout = -1;
+            readTimeout = -1;
+            connTimeout = -1;
+            shouldOptimizeTimeout = false;
+            return;
+        }
+
+        String timeout = "";
+        if (opType == AbfsRestOperationType.CreateFileSystem) {
```

Comment:
lets use switch case.

Reply:
Also, what if we have a field in AbfsRestOperationType enum containing the config name?

Reply:
Can try to have an enum with the AbfsRestOperationType and corresponding ConfigurationKey

Reply:
Resolving for future addition.